socketio/src/asynchronous/client/client.rs

```diff
@@ -26,49 +35,121 @@ use crate::{
 #[derive(Clone)]
 pub struct Client {
     /// The inner socket client to delegate the methods to.
-    socket: InnerSocket,
-    on: Arc<RwLock<HashMap<Event, Callback<DynAsyncCallback>>>>,
-    on_any: Arc<RwLock<Option<Callback<DynAsyncAnyCallback>>>>,
+    socket: Arc<RwLock<InnerSocket>>,
     outstanding_acks: Arc<RwLock<Vec<Ack>>>,
     // namespace, for multiplexing messages
     nsp: String,
     // Data send in the opening packet (commonly used as for auth)
     auth: Option<serde_json::Value>,
+    builder: Arc<RwLock<ClientBuilder>>,
+    manually_disconnected: Arc<AtomicBool>,
+    server_disconnected: Arc<AtomicBool>,
 }
 
 impl Client {
     /// Creates a socket with a certain address to connect to as well as a
     /// namespace. If `None` is passed in as namespace, the default namespace
     /// `"/"` is taken.
     /// ```
-    pub(crate) fn new<T: Into<String>>(
-        socket: InnerSocket,
-        namespace: T,
-        on: HashMap<Event, Callback<DynAsyncCallback>>,
-        on_any: Option<Callback<DynAsyncAnyCallback>>,
-        auth: Option<serde_json::Value>,
-    ) -> Result<Self> {
+    pub(crate) fn new(socket: InnerSocket, builder: ClientBuilder) -> Result<Self> {
         Ok(Client {
-            socket,
-            nsp: namespace.into(),
-            on: Arc::new(RwLock::new(on)),
-            on_any: Arc::new(RwLock::new(on_any)),
+            socket: Arc::new(RwLock::new(socket)),
+            nsp: builder.namespace.to_owned(),
             outstanding_acks: Arc::new(RwLock::new(Vec::new())),
-            auth,
+            auth: builder.auth.clone(),
+            builder: Arc::new(RwLock::new(builder)),
+            manually_disconnected: Arc::new(AtomicBool::new(false)),
+            server_disconnected: Arc::new(AtomicBool::new(false)),
         })
     }
 
     /// Connects the client to a server. Afterwards the `emit_*` methods can be
     /// called to interact with the server.
     pub(crate) async fn connect(&self) -> Result<()> {
         // Connect the underlying socket
-        self.socket.connect().await?;
+        self.socket.read().await.connect().await?;
 
         // construct the opening packet
         let auth = self.auth.as_ref().map(|data| data.to_string());
         let open_packet = Packet::new(PacketId::Connect, self.nsp.clone(), auth, None, 0, None);
 
-        self.socket.send(open_packet).await?;
+        self.socket.read().await.send(open_packet).await?;
+
+        Ok(())
+    }
+
+    pub(crate) async fn reconnect(&mut self) -> Result<()> {
+        let builder = self.builder.write().await;
+        let socket = builder.inner_create().await?;
+
+        // New inner socket that can be connected
+        let mut client_socket = self.socket.write().await;
+        *client_socket = socket;
```

Comment:
Lines 86-88, seem quite counterintuitive can you explain what/why you are doing this?

Reply:
I'm replacing `self.socket` with a new connection I've created on line 83. Since `connect()` needs to use it, I have to make sure I'm not holding on to a write lock before I call it on line 90.

Reply:
I've also added some comments now in the code below to clarify what this is doing.